src/platform/datapath_winuser.c

```diff
@@ -573,11 +747,27 @@ QuicDataPathInitialize(
         MaxProcCount = UINT16_MAX - 1;
     }
 
-    if (RecvCallback == NULL || UnreachableCallback == NULL || NewDataPath == NULL) {
+    if (NewDataPath == NULL) {
         Status = QUIC_STATUS_INVALID_PARAMETER;
         Datapath = NULL;
         goto Exit;
     }
+    if (UdpCallbacks != NULL) {
+        if (UdpCallbacks->Receive == NULL || UdpCallbacks->Unreachable == NULL) {
+            Status = QUIC_STATUS_INVALID_PARAMETER;
+            Datapath = NULL;
+            goto Exit;
+        }
+    }
+    if (TcpCallbacks != NULL) {
+        if (TcpCallbacks->Accept == NULL ||
+            TcpCallbacks->Connect == NULL ||
+            TcpCallbacks->Receive == NULL) {
+            Status = QUIC_STATUS_INVALID_PARAMETER;
+            Datapath = NULL;
+            goto Exit;
+        }
+    }
```

Comment:
enforce that specifying both Udp and Tcp callbacks should fail. A socket cannot be both TCP and UDP, right?

Reply:
This is the datapath layer, not socket layer. You can have one datapath handle that you use to create both UDP and TCP sockets.